Domains: Add advanced transfer options card

leonardost opened the pull request:
### Changes proposed in this Pull Request

**Note**: This depends on D70720-code.

This PR creates the advanced transfer options card in the new redesigned transfer page. This card contains a transfer lock toggle for the domain and a button to send the domain's authorization code to its owner - these actions were previously contained in the "Transfer to another registrar" option, which we won't have anymore.

A new action and reduce were created because the existing `DOMAIN_TRANSFER_CODE_REQUEST` one encompassed unlocking the domain and also disabling its privacy and sending the auth code via email, which will remain as separate actions now.

If the domain is currently in a transfer lock - either because it's a new domain (less than 60 days old) or because its contact info was updated and the owner chose not to opt-put of the lock - the toggle will be disabled and an explanation and a link to the relevant support docs will be shown to the user.

This is part of the domain management redesign project described in pcYYhz-m2-p2.

#### Screenshots

Old "Transfer to another registrar" page that will be removed

<img width="770" alt="Screen Shot 2021-11-26 at 15 11 35" src="https://user-images.githubusercontent.com/5324818/143618546-d68057f0-dc3a-4c39-83f4-84bafffbd033.png">

Advanced options card in desktop view

![Markup on 2021-12-08 at 16:04:22](https://user-images.githubusercontent.com/5324818/145268517-ec6f4aa7-c1d8-436c-9bdb-597c6025a0ae.png)

And in mobile view

![Markup on 2021-12-08 at 16:05:11](https://user-images.githubusercontent.com/5324818/145268525-9f518056-2a96-402b-8101-a400ec658c90.png)

When the domain is under a transfer lock, the following message will be displayed:

<img width="741" alt="Screen Shot 2021-12-08 at 16 03 34" src="https://user-images.githubusercontent.com/5324818/145269008-d9e860ad-7a35-47fc-8c30-95d7ab730f8a.png">

### Testing instructions

- Apply D70720-code to your sandbox if it hasn't been merged yet and sandbox `public-api`
- Build this branch locally or open the live Calypso link
- Go to "Upgrades > Domains"
- Select one of your registered domains or domain connections
- If you're in the live Calypso link, please append ?flags=domains/transfers-redesign to your URL to enable the appropriate feature flag
- Select the "Transfer your domain" options
- Please check that:
  - Advanced options card is present
  - Lock/unlock toggle works as expected
  - "Get authorization code" button works
  - Mobile view looks correct as shown in the screenshot above
  - If you selected a domain connection, ensure the advanced options do not appear, as they are only pertinent to registered domains
  - Also please check a domain which is under transfer lock

leonardost committed:
Create new domain transfer page

leonardost committed:
Make page's layout wide and the background white

leonardost committed:
Create new transfer item components

leonardost committed:
Remove index.jsx - it was replaced by the .tsx file

leonardost committed:
Refactor code and add advanced transfer options card

leonardost committed:
Create toggle labels

leonardost committed:
Add "send auth code" button functionality

a8ci18n commented on client/my-sites/domains/domain-management/transfer/transfer-page/index.tsx:
```diff
@@ -89,22 +99,87 @@ const TransferPage = ( props: TransferPageProps ): JSX.Element => {
+
+	const renderAdvancedTransferOptions = () => {
+		// translators: domain transfer lock
+		const disabledLockLabel = __( 'Transfer lock off' );
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **20** times:
`translate( 'Transfer lock: off' )` *ES Score: 12*
See 1 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/advanced-transfer-options-component)

a8ci18n commented on client/my-sites/domains/domain-management/transfer/transfer-page/index.tsx:
```diff
@@ -89,22 +99,87 @@ const TransferPage = ( props: TransferPageProps ): JSX.Element => {
+		// translators: domain transfer lock
+		const disabledLockLabel = __( 'Transfer lock off' );
+		// translators: domain transfer lock
+		const enabledLockLabel = __( 'Transfer lock on' );
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **23** times:
`translate( 'Transfer lock: on' )` *ES Score: 11*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/advanced-transfer-options-component)

a8ci18n commented on client/my-sites/domains/domain-management/transfer/transfer-page/index.tsx:
```diff
@@ -89,22 +99,87 @@ const TransferPage = ( props: TransferPageProps ): JSX.Element => {
+					) }
+				</p>
+				<Button primary={ false } busy={ isRequestingTransferCode } onClick={ requestTransferCode }>
+					{ __( 'Get authorization code' ) }
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **14** times:
`translate( 'Get domain authorization code' )` *ES Score: 7*
See 1 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/advanced-transfer-options-component)

leonardost committed:
Implement domain lock/unlock redux code

leonardost committed:
Create a new action for only locking a domain

The "cancel transfer out" action was used but it does other things
besides simply unlocking the domain. So I though it was better to
create a new action just for that.

leonardost committed:
Simplify some action handling method names

leonardost committed:
Query domain info on page load and rename variable that checks loading state

leonardost committed:
Reorder imports in alphabetical order

leonardost committed:
Add missing properties to props types declaration

leonardost committed:
Show toggle placeholder while loading domain info

leonardost committed:
Fix button loading property

leonardost committed:
Reorder properties declaration

leonardost committed:
Reorder properties declaration

leonardost committed:
Add specific function type declarations to props

leonardost committed:
Fix top margin of card in mobile view

leonardost committed:
Fix CSS differences with the design

leonardost committed:
Update error message when a problem occurs while locking/unlocking a domain

leonardost committed:
Add spaces after periods in error messages

leonardost committed:
Merge lock and unlock actions into one

leonardost committed:
Rename action DOMAIN_TRANFER_LOCK_TOGGLE to DOMAIN_TRANSFER_TOGGLE_LOCK

leonardost committed:
Do not show advanced transfer options if a domain connection is being managed

hambai reviewed (changes requested):
I believe we can simplify this PR and not keep all those new actions, action creators and reducers in the global state. We should be able to just use local component state for `isRequestingTransferCode` and `isLockingOrUnlockingDomain`. I also think that we can safely assume that if we got a success from the api endpoint we can just update the global state for the domain lock status and not make additional api call to get the updated status?

leonardost committed:
Remove DOMAIN_TRANSFER_CODE_ONLY_REQUEST action and related code

We don't need to use actions and redux for everything. This commit
removes that action and related code, stoing local state in the
component.

leonardost committed:
Remove DOMAIN_TRANSFER_TOGGLE_LOCK action and related code

leonardost committed:
Add DOMAIN_TRANSFER_UPDATE_LOCK action to update domains's transfer lock in global state

leonardost committed:
Fix style of "Transfer" header

leonardost committed:
Make methods with API calls async

leonardost committed:
Rename method `getDomainLockUnlockError` to `getDomainLockError`

leonardost committed:
Reorder const declaration

leonardost committed:
Reorder const declaration

leonardost committed:
Add alternative description when domain cannot be transfer locked

leonardost committed:
Remove unused code

leonardost committed:
Remove comment

leonardost committed:
Remove unused type declaration

leonardost commented:
Thanks for the review @hambai! You're definitely right, not everything needs to be in redux or in the global state. I started from the existing code and didn't even think about doing things more simply, so thanks for the discussion yesterday 🙇 

Now there's only one action to update the lock status of the domain when the API call is successful, so there's a lot less boilerplate 🙂 would love if you could take another look at this.

leonardost committed:
Remove unused variable

rafaelgallani reviewed (commented):
Nice job! I tested and all the visual changes were reflected, and the authorization code button also worked as expected. LGTM.

But there was a single point where I have noticed something weird:

![image](https://user-images.githubusercontent.com/18705930/144974915-89539c55-33f8-4a5f-b7f3-576846d401f8.png)

I wasn't able to turn off the transfer lock for my domain. 
I believe that's _probably_ because I marked the 60-day transfer lock when I previously updated (a few days ago) the domain's contact info. There's also the transfer's card message confirming it:

![image](https://user-images.githubusercontent.com/18705930/144975205-f7b3c600-2da6-4e7d-ba87-93925eea7525.png)

I think it might be a good idea to disable the button when the domain is in this state, otherwise, it might get some users confused.

leonardost commented:
Thanks for the review @rafaelgalani and nice catch! I'll add a check for the 60-day transfer lock 👍

leonardost committed:
Disable lock toggle if domain is in the 60-day transfer lock

leonardost committed:
Move translator comment to line above translated string

leonardost committed:
Add link to support docs if domain is in transfer lock

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/7030084

Thank you @leonardost for including a screenshot in the description! This is really helpful for our translators.

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/7030084).
